### PR TITLE
material-black-colors: init at 0-unstable-2020-12-17

### PR DESCRIPTION
**Pull request by D3vil0p3r**
material-black-colors: init at unstable-2020-12-17

image

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/ma/material-black-colors/package.nix [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out/share/icons
+    ${if colorVariants != [] then ''
+    name= cp -r \
+      ${lib.optionalString (colorVariants != []) toString colorVariants} \
+      $out/share/icons/
+    ''
+    else ''
+      cp -rf MB-Blueberry-Suru-GLOW MB-Cherry-Suru-GLOW MB-Lime-Suru-GLOW MB-Mango-Suru-GLOW MB-Pistachio-Suru-GLOW MB-Plum-Suru-GLOW Material-Black-Blueberry-Numix-FLAT Material-Black-Blueberry-Numix Material-Black-Blueberry-Suru Material-Black-Cherry-Numix-FLAT Material-Black-Cherry-Numix Material-Black-Cherry-Suru Material-Black-Lime-Numix-FLAT Material-Black-Lime-Numix Material-Black-Lime-Suru Material-Black-Mango-Numix-FLAT Material-Black-Mango-Numix Material-Black-Mango-Suru Material-Black-Pistachio-Numix-FLAT Material-Black-Pistachio-Numix Material-Black-Pistachio-Suru Material-Black-Plum-Numix-FLAT Material-Black-Plum-Numix Material-Black-Plum-Suru $out/share/icons/
```
Comment (SuperSandro2000) [inline pkgs/by-name/ma/material-black-colors/package.nix:33]: Can we dedupe this list with the one at the top and add some line breaks to short the line?

Comment (D3vil0p3r) [reply pkgs/by-name/ma/material-black-colors/package.nix:33]: > Can we dedupe this list with the one at the top and add some line breaks to short the line?

Dedupe the list how? Do you have an idea in mind?

Comment (SuperSandro2000) [reply pkgs/by-name/ma/material-black-colors/package.nix:33]: Not the list itself but the definition near the top of the file.

Comment (D3vil0p3r) [reply pkgs/by-name/ma/material-black-colors/package.nix:33]: > Not the list itself but the definition near the top of the file.

Something like this?
```nix
  iconset = "
    MB-Blueberry-Suru-GLOW \
    MB-Cherry-Suru-GLOW \
    MB-Lime-Suru-GLOW \
    MB-Mango-Suru-GLOW \
    MB-Pistachio-Suru-GLOW \
    MB-Plum-Suru-GLOW \
    Material-Black-Blueberry-Numix-FLAT \
    Material-Black-Blueberry-Numix \
    Material-Black-Blueberry-Suru \
    Material-Black-Cherry-Numix-FLAT \
    Material-Black-Cherry-Numix \
    Material-Black-Cherry-Suru \
    Material-Black-Lime-Numix-FLAT \
    Material-Black-Lime-Numix \
    Material-Black-Lime-Suru \
    Material-Black-Mango-Numix-FLAT \
    Material-Black-Mango-Numix \
    Material-Black-Mango-Suru \
    Material-Black-Pistachio-Numix-FLAT \
    Material-Black-Pistachio-Numix \
    Material-Black-Pistachio-Suru \
    Material-Black-Plum-Numix-FLAT \
    Material-Black-Plum-Numix \
    Material-Black-Plum-Suru \
  ";
...
    else ''
      cp -rf "${iconset}" $out/share/icons/
    ''
    }

    runHook postInstall
  '';
  ```

Comment (D3vil0p3r) [reply pkgs/by-name/ma/material-black-colors/package.nix:33]: @SuperSandro2000 what about this solution?
```nix
{ lib
, stdenvNoCC
, fetchFromGitHub
, colorVariants ? []
}:

let
  pname = "material-black-colors";
  colorVariantList = [
    "MB-Blueberry-Suru-GLOW"
    "MB-Cherry-Suru-GLOW"
    "MB-Lime-Suru-GLOW"
    "MB-Mango-Suru-GLOW"
    "MB-Pistachio-Suru-GLOW"
    "MB-Plum-Suru-GLOW"
    "Material-Black-Blueberry-Numix-FLAT"
    "Material-Black-Blueberry-Numix"
    "Material-Black-Blueberry-Suru"
    "Material-Black-Cherry-Numix-FLAT"
    "Material-Black-Cherry-Numix"
    "Material-Black-Cherry-Suru"
    "Material-Black-Lime-Numix-FLAT"
    "Material-Black-Lime-Numix"
    "Material-Black-Lime-Suru"
    "Material-Black-Mango-Numix-FLAT"
    "Material-Black-Mango-Numix"
    "Material-Black-Mango-Suru"
    "Material-Black-Pistachio-Numix-FLAT"
    "Material-Black-Pistachio-Numix"
    "Material-Black-Pistachio-Suru"
    "Material-Black-Plum-Numix-FLAT"
    "Material-Black-Plum-Numix"
    "Material-Black-Plum-Suru"
  ];

in
lib.checkListOfEnum "${pname}: color variants" colorVariantList colorVariants

stdenvNoCC.mkDerivation rec {
  inherit pname;
  version = "unstable-2020-12-17";

  src = fetchFromGitHub {
    owner = "rtlewis88";
    repo = "rtl88-Themes";
    rev = "3864d851aac7f4e76cf23717aee104de234aef74";
    hash = "sha256-BUJMd6Ltq16/HqqDbB5VDGIRSzLivXxNYZPT9sd6oTI=";
  };

  installPhase = ''
    runHook preInstall
    mkdir -p $out/share/icons
    ${if colorVariants != [] then ''
    name= cp -r \
      ${lib.optionalString (colorVariants != []) toString colorVariants} \
      $out/share/icons/
    ''
    else ''
      cp -rf ${lib.concatStringsSep " " colorVariantList} $out/share/icons/
    ''}
    runHook postInstall
  '';

  meta = with lib; {
    description = "Material Black Colors icons";
    homepage = "https://github.com/rtlewis88/rtl88-Themes/tree/material-black-COLORS";
    maintainers = with maintainers; [ d3vil0p3r ];
    platforms = platforms.all;
    license = licenses.gpl3Plus;
  };
}
```

Comment (SuperSandro2000) [reply pkgs/by-name/ma/material-black-colors/package.nix:33]: yes, like this 👍🏼

Comment (D3vil0p3r) [reply pkgs/by-name/ma/material-black-colors/package.nix]: > yes, like this 👍🏼

Fixed.

File: pkgs/by-name/ma/material-black-colors/package.nix [hunk 2]
```diff
@@ -0,0 +1,71 @@
+{ lib
+, stdenvNoCC
+, fetchFromGitHub
+, colorVariants ? []
```
[the remark this answers is not in the document]
Comment (D3vil0p3r) [reply pkgs/by-name/ma/material-black-colors/package.nix]: Fixed

File: pkgs/by-name/ma/material-black-colors/package.nix [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    "Material-Black-Plum-Numix"
+    "Material-Black-Plum-Suru"
+  ];
+
+in
+lib.checkListOfEnum "${pname}: color variants" colorVariantList colorVariants
+
+stdenvNoCC.mkDerivation {
+  inherit pname;
+  version = "unstable-2020-12-17";
```
Comment (h7x4) [inline pkgs/by-name/ma/material-black-colors/package.nix:41]: There's some relatively new versioning guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-naming

```suggestion
  version = "0-unstable-2020-12-17";
```

**Comment by h7x4** [comment]
Could you update the commit message as well, to reflect the new version tag?

**Comment by D3vil0p3r** [comment]
> Could you update the commit message as well, to reflect the new version tag?

@h7x4 fixed.

File: pkgs/by-name/ma/material-black-colors/package.nix [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    hash = "sha256-BUJMd6Ltq16/HqqDbB5VDGIRSzLivXxNYZPT9sd6oTI=";
+  };
+
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out/share/icons
+    ${if colorVariants != [] then ''
+      name= cp -r \
+        ${lib.optionalString (colorVariants != []) toString colorVariants} \
+        $out/share/icons/
```
[the remark this answers is not in the document]
Comment (D3vil0p3r) [reply pkgs/by-name/ma/material-black-colors/package.nix]: @h7x4 fixed as requested.

File: pkgs/by-name/ma/material-black-colors/package.nix [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out/share/icons
+    ${if colorVariants != [] then ''
+      cp -r ${lib.concatStringsSep " " colorVariants} $out/share/icons/
+    ''
+    else ''
+      cp -rf ${lib.concatStringsSep " " colorVariantList} $out/share/icons/
+    ''}
```
Comment (h7x4) [inline pkgs/by-name/ma/material-black-colors/package.nix:58]: Doing another take on this

```suggestion
    cp -r ${lib.concatStringsSep " " (if colorVariants != [] then colorVariants else colorVariantList)} $out/share/icons/
```

Comment (D3vil0p3r) [reply pkgs/by-name/ma/material-black-colors/package.nix]: @h7x4 fixed.

**Comment by h7x4** [comment]
@D3vil0p3r How long does this package spend in `fixupPhase` for you? For me, it spends an unreasonable amount of time, especially considering that it's not explicitly set.

**Comment by D3vil0p3r** [comment]
> @D3vil0p3r How long does this package spend in `fixupPhase` for you? For me, it spends an unreasonable amount of time, especially considering that it's not explicitly set.

29 minutes I see.

**Comment by h7x4** [comment]
bruh

File: pkgs/by-name/ma/material-black-colors/package.nix [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    hash = "sha256-BUJMd6Ltq16/HqqDbB5VDGIRSzLivXxNYZPT9sd6oTI=";
+  };
+
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out/share/icons
+    cp -r ${lib.concatStringsSep " " (if colorVariants != [] then colorVariants else colorVariantList)} $out/share/icons/
+    runHook postInstall
+  '';
+
```
Comment (h7x4) [inline pkgs/by-name/ma/material-black-colors/package.nix:56]:
```suggestion

  dontFixup = true;

```

Comment (D3vil0p3r) [reply pkgs/by-name/ma/material-black-colors/package.nix:56]: @h7x4 applied fix.

**Review by h7x4 — APPROVED** [review]

**Pull request merged**